### ISSUE TITLE
Bug: no-useless-assignment false positive for variables used in parameter decorator arguments

### ISSUE THREAD

**Issue by ropdias** [eslint/eslint#20947]
### Environment

Node version: v24.14.0
npm version: v11.9.0
Local ESLint version: v10.4.1 (Currently used)
Global ESLint version: Not found
Operating System: linux 6.6.87.2-microsoft-standard-WSL2

### What parser are you using?

@typescript-eslint/parser

### What did you do?

<details>
<summary>Configuration</summary>

```js
import eslint from '@eslint/js'
import { defineConfig } from 'eslint/config'
import tseslint from 'typescript-eslint'

export default defineConfig(
  eslint.configs.recommended,
  ...tseslint.configs.recommendedTypeChecked,
  {
    languageOptions: {
      parserOptions: {
        projectService: true,
      },
    },
    rules: {
      'no-useless-assignment': 'error',
    },
  },
)
```
</details>

```js
function Body(_pipe: unknown) {
  return function (
    _target: unknown,
    _propertyKey: string,
    _parameterIndex: number,
  ) {}
}

const pipe = {}

class Controller {
  handler(@Body(pipe) _body: unknown) {}
}
```

### What did you expect to happen?

No lint error.

`pipe` is read and passed as an argument to the parameter decorator expression.

### What actually happened?

`no-useless-assignment` reports:

```txt
The value assigned to 'pipe' is not used in subsequent statements.
```

### Link to Minimal Reproducible Example

https://eslint.org/play/#eyJ0ZXh0IjoiZnVuY3Rpb24gQm9keShfcGlwZTogdW5rbm93bikge1xuICByZXR1cm4gZnVuY3Rpb24gKFxuICAgIF90YXJnZXQ6IHVua25vd24sXG4gICAgX3Byb3BlcnR5S2V5OiBzdHJpbmcsXG4gICAgX3BhcmFtZXRlckluZGV4OiBudW1iZXIsXG4gICkge31cbn1cblxuY29uc3QgcGlwZSA9IHt9XG5cbmNsYXNzIENvbnRyb2xsZXIge1xuICBoYW5kbGVyKEBCb2R5KHBpcGUpIF9ib2R5OiB1bmtub3duKSB7fVxufSIsIm9wdGlvbnMiOnsicnVsZXMiOnsibm8tdXNlbGVzcy1hc3NpZ25tZW50IjpbImVycm9yIl19LCJsYW5ndWFnZU9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hRmVhdHVyZXMiOnsianN4Ijp0cnVlfX0sInBhcnNlciI6IkB0eXBlc2NyaXB0LWVzbGludC9wYXJzZXIiLCJlY21hVmVyc2lvbiI6ImxhdGVzdCIsInNvdXJjZVR5cGUiOiJtb2R1bGUifX19

### Participation

- [ ] I am willing to submit a pull request for this issue.

### AI acknowledgment

- [ ] I did not use AI to generate this issue report.
- [x] (If the above is not checked) I have reviewed the AI-generated content before submitting.

### Additional comments

This appears to specifically affect reads inside parameter decorator expressions and reproduces in the official ESLint playground using `@typescript-eslint/parser` with no external framework dependencies.

Equivalent patterns using class decorators or method decorators do not reproduce the issue.

This pattern is commonly used in frameworks like NestJS:
```ts
@Body(validationPipe)
```
where module-level pipe instances are passed into parameter decorators.

**Comment by DMartens** [comment]
While I can reproduce this problem, we only support [stage 4](https://github.com/eslint/eslint#what-about-experimental-features) proposal and decorators (which do not include parameter decorators) is at stage 3.
Further `typescript-eslint` also does not support parameter decorators as TypeScript does not [allow them](https://typescript-eslint.io/play/#ts=5.9.2&fileType=.tsx&code=GYVwdgxgLglg9mABAITgEwJ4AoD6AHGPAUwC5FwBrMOAdzAEpEBvAKEUQCcioQOlRIsBIixt2iHFACGHAObcylanQA0Y9vg5xiHKBgDSRDGQDOUDjDCy14iXhlSAttyIcAkmDREAHmTAhHACNXG0RGJkQAXxZolggEM0QCYkQAXmZYiAAbKRMTRABhBHM4LKzXZjEACylPco4sAAFUTCxkokYcQPRjcjAqWgZmKJigA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false).

**Comment by kirkwaiblinger** [comment]
> Further `typescript-eslint` also does not support parameter decorators as TypeScript does not [allow them](https://typescript-eslint.io/play/#ts=5.9.2&fileType=.tsx&code=GYVwdgxgLglg9mABAITgEwJ4AoD6AHGPAUwC5FwBrMOAdzAEpEBvAKEUQCcioQOlRIsBIixt2iHFACGHAObcylanQA0Y9vg5xiHKBgDSRDGQDOUDjDCy14iXhlSAttyIcAkmDREAHmTAhHACNXG0RGJkQAXxZolggEM0QCYkQAXmZYiAAbKRMTRABhBHM4LKzXZjEACylPco4sAAFUTCxkokYcQPRjcjAqWgZmKJigA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false).

This isn't quite right; this [does work](https://typescript-eslint.io/play/#ts=5.9.2&fileType=.tsx&code=GYVwdgxgLglg9mABAITgEwJ4AoD6AHGPAUwC5FwBrMOAdzAEpEBvAKEUQCcioQOlRIsBIixt2iHFACGHAObcylanQA0Y9vg5xiHKBgDSRDGQDOUDjDCy14iXhlSAttyIcAkmDREAHmTAhHACNXG0RGJkQAXxZolggEM0QCYkQAXmZYiAAbKRMTRABhBHM4LKzXZjEACylPco4sAAFUTCxkokYcQPRjcjAqWgZmKJigA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3QAacDUjoAHoizJ46IngCGqACL9YmNXl3Cwo9FIC%2BIM0A&tokens=false) with [experimentalDecorators: true](https://www.typescriptlang.org/tsconfig/#experimentalDecorators)

**Comment by ropdias** [comment]
Thank you @kirkwaiblinger for the correction! Indeed, with `experimentalDecorators: true`, this is fully supported by TypeScript and is a standard pattern heavily used in frameworks like NestJS.

Since the ESLint core scope rules focus strictly on standardized TC39 ECMAScript features (which do not include parameter decorators), it makes total sense that this won't be handled here. 

The root cause is likely inside the `typescript-eslint` scope manager not registering references inside legacy parameter decorators as a "read" operation for core rules.

I have opened a tracking issue on their repository so they can evaluate an AST/scope adjustment or a TS-aware extension rule. Link to the issue: [https://github.com/typescript-eslint/typescript-eslint/issues/12407](url)

Thanks everyone for the insights!
